### PR TITLE
podman: add Linux support

### PR DESCRIPTION
**Pull request by branchv**
Created with `brew bump-formula-pr`.

~`resource` blocks may require updates.~

### CODE REVIEW

File: Formula/podman.rb [hunk 1]
```diff
@@ -36,6 +36,8 @@ def install
     end
     on_linux do
       bin.install "bin/podman-remote"
+      system "make", "podman"
+      bin.install "bin/podman"
```
Comment (branchvincent) [inline Formula/podman.rb:40]: adding this after following the discussion from https://github.com/Homebrew/homebrew-core/pull/83578#issuecomment-902126237,

File: Formula/podman.rb [hunk 2]
```diff
@@ -29,6 +48,20 @@ def install
       "linux"
     end
 
+    # Install podman, only on Linux
+    on_linux do
+      resource("cni-plugins").stage do
```
Comment (branchvincent) [inline Formula/podman.rb:53]: these were the minimum plugins I needed for `podman` to be happy, should these be moved to a separate formula?

File: Formula/podman.rb [hunk 3]
```diff
@@ -17,6 +17,25 @@ class Podman < Formula
...
+    depends_on "libseccomp"
+  end
+
+  on_linux do
+    resource "cni-plugins" do
+      url "https://github.com/containernetworking/plugins/archive/v1.0.0.tar.gz"
+      sha256 "8331f2c09d13012baacdec4d0ce07c081d184354a6c4e5aba40584a88a53437f"
+    end
+
+    resource "containers.conf" do
```
Comment (branchvincent) [inline Formula/podman.rb:33]: We need this to override the default `cni_plugin_dirs` so `podman` can find our plugins. I was hoping we could install our own so there's no manual configuration needed, but the only filepaths searched are https://github.com/containers/common/blob/6143ffe54e5f9da2f69a4d231c1123af07fdc1e4/pkg/config/containers.conf#L11-L13 which i don't think we can write to?

**Comment by branchvincent** [comment]
tests failed because it is running in docker without `--privileged`

```
cannot clone: Operation not permitted
Error: cannot re-exec process
```

**Comment by SMillerDev** [comment]
Maybe we can base the test on https://developers.redhat.com/blog/2020/09/25/rootless-containers-with-podman-the-basics#example__using_rootless_containers or just expect a failure?

**Comment by danielnachun** [comment]
I think the only thing we are really missing here is to fix the test.  @branchvincent we just need to catch the right error here.  I also think that `cni-plugins` should get its own formula at some point, as it is packaged separately by other distros.

**Comment by branchvincent** [comment]
@danielnachun i'll update this to expect the right failure.

do you by chance have a linux machine to test this? i was hoping to be more confident that the formula works, but I ran into other issues testing in Docker

**Comment by danielnachun** [comment]
Actually I think I do have one I could test it on.  Testing these virtualization related packages is tricky because we're already running inside a VM and nested virtualization is not trivial.

**Comment by danielnachun** [comment]
Let me try this now on my Linux machine and see if it works.  I'll post what the test output is on there.

File: Formula/podman.rb (hunk 2 above)
Comment (afbjorklund) [reply Formula/podman.rb:55]: Most likely

File: Formula/podman.rb (hunk 3 above)
Comment (afbjorklund) [reply Formula/podman.rb]: Usually patch it in the source code (so that no configuration is needed), or add symlinks to the default locations.

**Comment by afbjorklund** [comment]
@branchvincent : I tried the linux version on Ubuntu, after rebasing. I'm not sure how homebrew normally handles systemd, so I'm not sure if the podman.socket will be supported (it is needed for API access etc).

Copying things to `/etc` sounds like a hack, and the CNI plugins should not be dumped in the top-level `libexec` (they should use a subdirectory, preferrably a separate port rather than being bundled with podman)

Note that you also need to patch/config the location of `/etc/cni/net.d`, for the network.

There are also a bunch of other **required** config files, hiding in containers-common.

**Comment by afbjorklund** [comment]
Note that this "libexec" issue is basically the same problem as in: https://github.com/Homebrew/homebrew-core/pull/90379

The "gvproxy" program should probably have been in a `podman` subdirectory.

```toml
# A is a list of directories which are used to search for helper binaries.
#
#helper_binaries_dir = [
#  "/usr/local/libexec/podman",
#  "/usr/local/lib/podman",
#  "/usr/libexec/podman",
#  "/usr/lib/podman",
#]
```

Just like "catatonit" (like: docker-init), and "conmon" (like: containerd)

```toml
# Container init binary, if init=true, this is the init binary to be used for containers.
#
#init_path = "/usr/libexec/podman/catatonit"
```

```toml
# Paths to look for the conmon container manager binary
#
#conmon_path = [
#  "/usr/libexec/podman/conmon",
#  "/usr/local/libexec/podman/conmon",
#  "/usr/local/lib/podman/conmon",
#  "/usr/bin/conmon",
#  "/usr/sbin/conmon",
#  "/usr/local/bin/conmon",
#  "/usr/local/sbin/conmon"
#]
```

Just that CNI can (theoretically) be shared with other programs, too.

**Comment by danielnachun** [comment]
For some reference, Arch Linux packages `cni-plugins` and `containers-common` as separate packages.  That makes me think we need separate formulae for those as well.

Additionally, Arch Linux includes `fuse-overlayfs`, `slirp4netns` and `iptables` as additional dependencies.  We have formulae for the first 2 but not `iptables`, which could also warrant a new formula.  There are also a number of optional dependencies whose importance I'm not sure of yet: https://archlinux.org/packages/community/x86_64/podman/.

I agree that copying things to `etc` isn't ideal.  It seems like we need to set alternate paths in some config files to point `podman` towards the formula `optdir` instead of `/etc` - hopefully nothing is hard-coded.

**Comment by afbjorklund** [comment]
> hopefully nothing is hard-coded

But **everything** is hard-coded, so you will need to patch _something_ (either the values, or the location of the config)


```go
        // _configPath is the path to the containers/containers.conf
        // inside a given config directory.
        _configPath = "containers/containers.conf"
        // DefaultContainersConfig holds the default containers config path
        DefaultContainersConfig = "/usr/share/" + _configPath
        // OverrideContainersConfig holds the default config path overridden by the root user
        OverrideContainersConfig = "/etc/" + _configPath
```

Might be better to patch the default values, and leave the default containers configuration up to the local overrides ?

```go
        // It may seem a bit unconventional, but it is necessary to do so
        DefaultCNIPluginDirs = []string{
                "/usr/local/libexec/cni",
                "/usr/libexec/cni",
                "/usr/local/lib/cni",
                "/usr/lib/cni",
                "/opt/cni/bin",
        }
```
```go
        // CNIConfigDir is the path where CNI config files exist
        CNIConfigDir = "/etc/cni/net.d"
```

**Comment by danielnachun** [comment]
I think patching the default values makes sense.  I wonder how difficult it would be to make an upstream PR to provide a way to add directories to the default values through a build time flag?  That would be the ideal outcome.

**Pull request closed without merging**